EE-106: Parsing methods return Error instead of panic!

goral09 opened the pull request:
## Overview
Instead of panic!-ing when the ipc object can't be parsed into a domain one, we return `Result::Err`. Any parsing errors short circuits the execution and returns the error to the node.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://casperlabs.atlassian.net/browse/EE-106

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [ ] This PR meets [CasperLabs development coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [ ] If this PR adds a new feature, this PR includes tests related to this feature.
- [ ] You assigned one person to review this PR

### If you are not a member of the core development team, please confirm:
- [ ] You signed the commit. Merging requires a signature. Please see the [Signing Commits](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/4390963/Signing+Commits) for instructions.
- [ ] Your GitHub account is also an account with [Drone CI](http://3.16.200.31/). Unit tests will not run on your PR unless you have an account with Drone. Merge requires passed unit tests.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

birchmd commented on execution-engine/comm/src/engine_server/mappings.rs:
```diff
@@ -1,24 +1,41 @@
 /// Helper method for turning instances of Value into Transform::Write.
-fn transform_write(v: common::value::Value) -> storage::transform::Transform {
-    storage::transform::Transform::Write(v)
+fn transform_write(v: common::value::Value) -> Result<storage::transform::Transform, ParsingError> {
```
Could use `ParseResult` here

birchmd commented on execution-engine/comm/src/engine_server/mappings.rs:
```diff
@@ -1,24 +1,41 @@
+
+impl TryFrom<&super::ipc::Transform> for storage::transform::Transform {
+    type Error = ParsingError;
+    fn try_from(tr: &super::ipc::Transform) -> Result<storage::transform::Transform, ParsingError> {
```
Could use `ParseResult` here

birchmd commented on execution-engine/comm/src/engine_server/mappings.rs:
```diff
@@ -1,24 +1,41 @@
+            Ok(storage::transform::Transform::Identity)
         } else if tr.has_add_keys() {
-            let keys_map = tr
+            let keys_map: BTreeMap<String, common::key::Key> = tr
```
Is this type annotation needed, given the one on the `collect`?

goral09 replied:
No, it isn't but the one on the `collect` is (and it doesn't work if I removed the latter).

birchmd commented on execution-engine/comm/src/engine_server/mappings.rs:
```diff
@@ -1,24 +1,41 @@
+                    let local_nk = nk.clone();
+                    local_nk.get_key().try_into().map(|k| (local_nk.name, k))
+                })
+                .collect::<Result<BTreeMap<String, common::key::Key>, ParsingError>>()?;
```
Could use `ParseResult` here

birchmd commented on execution-engine/comm/src/engine_server/mappings.rs:
```diff
@@ -157,13 +180,16 @@ impl From<storage::transform::Transform> for super::ipc::Transform {
-    fn from(from: &[super::ipc::NamedKey]) -> Self {
+impl TryFrom<&[super::ipc::NamedKey]> for URefMap {
+    type Error = ParsingError;
+    fn try_from(from: &[super::ipc::NamedKey]) -> Result<Self, ParsingError> {
```
Could use `ParseResult` here

birchmd commented on execution-engine/comm/src/engine_server/mappings.rs:
```diff
@@ -157,13 +180,16 @@ impl From<storage::transform::Transform> for super::ipc::Transform {
         let mut tree: BTreeMap<String, common::key::Key> = BTreeMap::new();
-        for nk in from {
-            let _ = tree.insert(nk.get_name().to_string(), nk.get_key().into());
+        for mut nk in from {
```
Why does `nk` need the `mut` modifier now?

goral09 replied:
It doesn't. It is a leftover of me waiting with the compiler. Nice catch.

birchmd commented on execution-engine/comm/src/engine_server/mappings.rs:
```diff
@@ -206,23 +232,25 @@ impl From<&common::key::Key> for super::ipc::Key {
+impl TryFrom<&super::ipc::Key> for common::key::Key {
+    type Error = ParsingError;
+
+    fn try_from(ipc_key: &super::ipc::Key) -> Result<Self, ParsingError> {
```
Could use `ParseResult` here

birchmd commented on execution-engine/comm/src/engine_server/mappings.rs:
```diff
@@ -244,17 +272,19 @@ impl From<storage::op::Op> for super::ipc::Op {
-    fn from(from: &super::ipc::TransformEntry) -> Self {
+impl TryFrom<&super::ipc::TransformEntry> for (common::key::Key, storage::transform::Transform) {
+    type Error = ParsingError;
+    fn try_from(from: &super::ipc::TransformEntry) -> Result<Self, ParsingError> {
```
Could use `ParseResult` here

birchmd commented on execution-engine/comm/src/engine_server/mod.rs:
```diff
@@ -103,9 +116,25 @@ impl<R: DbReader, H: History<R>> ipc_grpc::ExecutionEngineService for EngineStat
-        let result = apply_effect_result_to_ipc(self.apply_effect(prestate_hash, effects));
-        grpc::SingleResponse::completed(result)
+        let effects_result: Result<HashMap<Key, Transform>, ParsingError> = p
+            .clone()
```
Why is it necessary to `clone` `p` now?

goral09 replied:
It isn't. Another leftover.

goral09 commented:
@birchmd I've updated the PR (pushed with force :muscle:). I've removed the `ParsingResult` entirely as it would be mostly used in a return type of the `TryFrom` 's method and I left the type from the trait for clarity.
